core/src/test/java/kafka/security/minikdc/MiniKdc.java

```diff
@@ -0,0 +1,513 @@
+            });
+        }).collect(Collectors.toList());
+        keytab.setEntries(entries);
+        keytab.write(keytabFile);
+    }
+
+    private void initDirectoryService() throws Exception {
+        ds = new DefaultDirectoryService();
+        ds.setInstanceLayout(new InstanceLayout(workDir));
+        ds.setCacheService(new CacheService());
+
+        // first load the schema
+        InstanceLayout instanceLayout = ds.getInstanceLayout();
+        File schemaPartitionDirectory = new File(instanceLayout.getPartitionsDirectory(), "schema");
+        DefaultSchemaLdifExtractor extractor = new DefaultSchemaLdifExtractor(instanceLayout.getPartitionsDirectory());
+        extractor.extractOrCopy();
+
+        LdifSchemaLoader loader = new LdifSchemaLoader(schemaPartitionDirectory);
+        DefaultSchemaManager schemaManager = new DefaultSchemaManager(loader);
+        schemaManager.loadAllEnabled();
+        ds.setSchemaManager(schemaManager);
+
+        // Init the LdifPartition with schema
+        LdifPartition schemaLdifPartition = new LdifPartition(schemaManager, ds.getDnFactory());
+        schemaLdifPartition.setPartitionPath(schemaPartitionDirectory.toURI());
+
+        // The schema partition
+        SchemaPartition schemaPartition = new SchemaPartition(schemaManager);
+        schemaPartition.setWrappedPartition(schemaLdifPartition);
+        ds.setSchemaPartition(schemaPartition);
+
+        JdbmPartition systemPartition = new JdbmPartition(ds.getSchemaManager(), ds.getDnFactory());
+        systemPartition.setId("system");
+        systemPartition.setPartitionPath(new File(ds.getInstanceLayout().getPartitionsDirectory(), systemPartition.getId()).toURI());
+        systemPartition.setSuffixDn(new Dn(ServerDNConstants.SYSTEM_DN));
+        systemPartition.setSchemaManager(ds.getSchemaManager());
+        ds.setSystemPartition(systemPartition);
+
+        ds.getChangeLog().setEnabled(false);
+        ds.setDenormalizeOpAttrsEnabled(true);
+        ds.addLast(new KeyDerivationInterceptor());
+
+        // create one partition
+        String orgName = config.getProperty(ORG_NAME).toLowerCase(Locale.ENGLISH);
+        String orgDomain = config.getProperty(ORG_DOMAIN).toLowerCase(Locale.ENGLISH);
+        JdbmPartition partition = new JdbmPartition(ds.getSchemaManager(), ds.getDnFactory());
+        partition.setId(orgName);
+        partition.setPartitionPath(new File(ds.getInstanceLayout().getPartitionsDirectory(), orgName).toURI());
+        Dn dn = new Dn("dc=" + orgName + ",dc=" + orgDomain);
+        partition.setSuffixDn(dn);
+        ds.addPartition(partition);
+
+        // indexes
+        Set<Index<?, String>> indexedAttributes = new HashSet<>(
+                Arrays.asList(
+                        new JdbmIndex<>("objectClass", false),
+                        new JdbmIndex<>("dc", false),
+                        new JdbmIndex<>("ou", false)));
```

Comment:
We could simply use `put` to avoid the secondary iteration since the entries are known.